File: Java/Waiter.java
// Производный класс Waiter наследуется от Employee и реализует множественное наследование
// (от абстрактного класса Employee и интерфейсов Payable, Displayable)
public class Waiter extends Employee {
    private int tablesServed;
    private double tips;
    private ServingStrategy servingStrategy;

    public Waiter(String name, double hourlyRate, int tablesServed) {
        this(name, hourlyRate, tablesServed, resolveStrategy());
    }

    public Waiter(String name, double hourlyRate, int tablesServed, ServingStrategy strategy) {
        super(name, "Waiter", hourlyRate); // Вызов конструктора базового класса
        this.tablesServed = tablesServed;
        this.tips = 0.0;
        this.servingStrategy = strategy != null ? strategy : resolveStrategy();
    }

    public Waiter(String name, double hourlyRate) {
        this(name, hourlyRate, 0);
    }

    // Переопределение метода базового класса с вызовом базового метода
    @Override
    public void showInfo() {
        super.showInfo(); // Вызов метода базового класса
        System.out.println("  Tables Served: " + tablesServed);
        System.out.println("  Tips: " + tips);
        System.out.println("  Serving strategy: " + (servingStrategy != null ? servingStrategy.getName() : "none"));
        System.out.println("  Total Earnings: " + (calculateSalary() + tips));
    }

    // Переопределение метода без вызова базового метода
    @Override
    public double calculateSalary() {
        // Официанты получают базовую зарплату плюс бонус за обслуженные столы
        double baseSalary = hoursWorked * hourlyRate;
        double bonus = servingStrategy != null
                ? servingStrategy.computeBonus(hoursWorked, hourlyRate, tablesServed, tips)
                : tablesServed * 50.0; // Бонус 50 за стол
        return baseSalary + bonus;
    }

    @Override
    public String getEmployeeType() {
        return "Waiter";
    }

    // Новые методы
    public void serveTable(int tableNumber) {
        tablesServed++;
    }

    public void addTips(double amount) {
        if (amount > 0.0) {
            tips += amount;
        }
    }

    public int getTablesServed() {
        return tablesServed;
    }

    public double getTips() {
        return tips;
    }

    public void setServingStrategy(ServingStrategy servingStrategy) {
        this.servingStrategy = servingStrategy;
    }

    public String getServingStrategyName() {
        return servingStrategy != null ? servingStrategy.getName() : "none";
    }

    // Переопределение метода интерфейса Payable
    @Override
    public double getAmount() {
        return calculateSalary() + tips;
    }

    @Override
    public String getPaymentInfo() {
        return "Waiter " + name + " total earnings: " + getAmount() + " (salary: " + 
               calculateSalary() + ", tips: " + tips + ")";
    }

    private static ServingStrategy resolveStrategy() {
        String mode = System.getProperty("waiter.mode");
        if (mode == null || mode.isEmpty()) {
            mode = System.getenv("WAITER_MODE");
        }
        if ("event".equalsIgnoreCase(mode)) {
            return new EventServingStrategy();
        }
        return new StandardServingStrategy();
    }
}

File: Waiter.cpp
#include "Waiter.hpp"
#include <cstdlib>
#include <iostream>
#include <memory>
#include <string>

namespace {
std::shared_ptr<ServingPolicy> createPolicyFromEnv() {
    const char* mode = std::getenv("WAITER_POLICY");
    std::string value = (mode != nullptr) ? std::string(mode) : "";
    if (value == "event") {
        return std::make_shared<EventServingPolicy>();
    }
    return std::make_shared<StandardServingPolicy>();
}
} // namespace

Waiter::Waiter(int id, const std::string& name, double rate, int tablesServed)
    : Employee(id, name, "Waiter", rate),
      tablesServed(tablesServed),
      tips(0.0),
      servingPolicy(createPolicyFromEnv())
{
    // Вызов конструктора базового класса Employee в списке инициализации
}

Waiter::Waiter(int id,
               const std::string& name,
               double rate,
               std::shared_ptr<ServingPolicy> policy,
               int tablesServed)
    : Employee(id, name, "Waiter", rate),
      tablesServed(tablesServed),
      tips(0.0),
      servingPolicy(policy ? policy : createPolicyFromEnv()) {}

Waiter::Waiter(const Waiter& other)
    : Employee(other),
      tablesServed(other.tablesServed),
      tips(other.tips),
      servingPolicy(other.servingPolicy)
{
    // Вызов конструктора копирования базового класса
}

void Waiter::showInfo() const {
    // Переопределение метода с вызовом базового метода
    Employee::showInfo(); // Вызов метода базового класса
    std::cout << "  Tables Served: " << this->tablesServed << std::endl;
    std::cout << "  Tips: " << this->tips << std::endl;
    std::cout << "  Serving policy: " << this->getServingPolicyName() << std::endl;
    std::cout << "  Total Earnings: " << (this->calculateSalary() + this->tips) << std::endl;
}

double Waiter::calculateSalary() const {
    // Переопределение без вызова базового метода
    // Официанты получают базовую зарплату плюс бонус за обслуженные столы
    double baseSalary = this->hoursWorked * this->hourlyRate;
    double bonus = servingPolicy ? servingPolicy->computeBonus(this->hoursWorked,
                                                               this->hourlyRate,
                                                               this->tablesServed,
                                                               this->tips)
                                 : this->tablesServed * 50.0;
    return baseSalary + bonus;
}

void Waiter::serveTable(int tableNumber) {
    this->tablesServed++;
}

void Waiter::addTips(double amount) {
    if (amount > 0.0) {
        this->tips += amount;
    }
}

Waiter& Waiter::operator=(const Waiter& other) {
    if (this != &other) {
        Employee::operator=(other); // Вызов оператора базового класса
        this->tablesServed = other.tablesServed;
        this->tips = other.tips;
        this->servingPolicy = other.servingPolicy;
    }
    return *this;
}

Waiter& Waiter::operator=(const Employee& other) {
    Employee::operator=(other); // Присваивание только базовой части
    // Поля, специфичные для Waiter, остаются неизменными
    return *this;
}

